src/webgl/shader.js

```diff
@@ -16,5 +19,286 @@ module.exports = {
+    attribute.location = location;
+    attribute.type = attributeInfo.type;
+    attribute.size = attributeInfo.size;
+    this.attributes[name] = attribute;
+  }
+
+  this.loadedAttributes = true;
+};
+
+p5.Shader.prototype.loadUniforms = function() {
+  if (this.loadedUniforms) {
+    return;
+  }
+
+  var gl = this._renderer.GL;
+
+  // Inspect shader and cache uniform info
+  var numUniforms = gl.getProgramParameter(this.glProgram,
+    gl.ACTIVE_UNIFORMS);
+
+  var samplerIndex = 0;
+  for(var i = 0; i < numUniforms; ++i){
+    var uniformInfo = gl.getActiveUniform(this.glProgram, i);
+    var uniform = {};
+    uniform.location = gl.getUniformLocation(this.glProgram, uniformInfo.name);
+    uniform.size = uniformInfo.size;
+    var uniformName = uniformInfo.name;
+    //uniforms thats are arrays have their name returned as
+    //someUniform[0] which is a bit silly so we trim it
+    //off here. The size property tells us that its an array
+    //so we dont lose any information by doing this
+    if(uniformInfo.size > 1) {
+      uniformName = uniformName.substring(0,
+        uniformName.indexOf('[0]'));
+    }
+    uniform.name = uniformName;
+    uniform.type = uniformInfo.type;
+    if(uniform.type === gl.SAMPLER_2D) {
+      uniform.samplerIndex = samplerIndex;
+      samplerIndex++;
+    }
+    this.uniforms[uniformName] = uniform;
+  }
+  this.loadedUniforms = true;
+};
+
+p5.Shader.prototype.compile = function() {
+  // TODO
+};
+
+// TODO: check context
+p5.Shader.prototype.bind = function () {
+  this.init();
+  if (!this.bound) {
+    this.useProgram();
+    this.bound = true;
+    this.bindTextures();
+
+    this.loadAttributes();
+    this.loadUniforms();
+
+    this._renderer._setDefaultCamera();
+    this._setMatrixUniforms();
+  }
+};
+
+p5.Shader.prototype.unbind = function () {
+  if (this.bound) {
+    this.unbindTextures();
+    //this._renderer.GL.useProgram(0); ??
+    this.bound = false;
+  }
+};
+
+p5.Shader.prototype.bindTextures = function () {
+  // TODO: migrate stuff from material.js here
+  // - OR - have material.js define this function
+};
+
+p5.Shader.prototype.unbindTextures = function () {
+  // TODO: migrate stuff from material.js here
+  // - OR - have material.js define this function
+};
+
+p5.Shader.prototype._setMatrixUniforms = function() {
+  this.setUniform('uProjectionMatrix', this._renderer.uPMatrix.mat4);
+  this.setUniform('uModelViewMatrix', this._renderer.uMVMatrix.mat4);
+
+  this._renderer.uNMatrix.inverseTranspose(this._renderer.uMVMatrix);
+  this.setUniform('uNormalMatrix', this._renderer.uNMatrix.mat3);
+};
+
+p5.Shader.prototype.useProgram = function () {
+  var gl = this._renderer.GL;
+  gl.useProgram(this.glProgram);
+};
+
+/**
+ * Wrapper around gl.uniform functions.
+ * As we store uniform info in the shader we can use that
+ * to do type checking on the supplied data and call
+ * the appropriate function.
+ */
+p5.Shader.prototype.setUniform = function(uniformName, data)
```

Comment:
Yeah I hit that when making the tests for loadPixels(). From what I can tell it isn't possible to run those WebGL unit tests from the cmd line build process (I could very well be wrong!) but if you want to run those tests you can run the browser version of the build tests with the included html file. That is how I ran my tests.